app-backend/common/src/main/scala/STS.scala

```diff
@@ -0,0 +1,83 @@
+package com.azavea.rf.common
+
+import java.sql.Timestamp
+
+import com.amazonaws.regions.Regions
+import com.amazonaws.services.securitytoken.AWSSecurityTokenServiceClientBuilder
+import com.amazonaws.services.securitytoken.model.{GetFederationTokenRequest, Credentials => STSCredentials}
+
+package object STS {
+
+  case class Credentials (
+    AccessKeyId: String,
+    SecretAccessKey: String,
+    SessionToken: String,
+    Expiration: Timestamp
+  )
+
+  def toCredentials(result: STSCredentials): Credentials =
+    Credentials(
+      result.getAccessKeyId,
+      result.getSecretAccessKey,
+      result.getSessionToken,
+      new Timestamp(result.getExpiration.getTime)
+    )
+
+  def getCredentialsForUpload(uploadId: String, bucket: String): Credentials =
+    getCredentialsForUpload(uploadId, bucket, "user-uploads")
+
+  def getCredentialsForUpload(uploadId: String, bucket: String, path: String): Credentials =
+    getCredentialsForUpload(uploadId, bucket, path, Regions.US_EAST_1)
+
+  def getCredentialsForUpload(uploadId: String, bucket: String, path: String, regions: Regions): Credentials = {
+    val client = AWSSecurityTokenServiceClientBuilder.standard.withRegion(regions).build
+
+    // Federation user's name must be less than or equal to 32 characters.
+    // We set it to the uploadId uuid without hyphens, which is exactly 32 characters.
+    // We also cap it to 32 to be safe.
+    val userName = uploadId.filter(_ != '-').take(32)
+
+    // We give the Federation user ability to list objects within their folder,
+    // and upload, download, and delete objects within their folder.
+    // Policy must be less than or equal to 2048 bytes, so we remove all whitespace.
+    val policy =
+      s"""
+         |{
+         |  "Version": "2012-10-17",
+         |  "Statement": [
+         |    {
+         |      "Effect": "Allow",
+         |      "Action": [
+         |        "s3:ListBucket"
+         |      ],
+         |      "Resource": [
+         |        "arn:aws:s3:::${bucket}"
+         |      ],
+         |      "Condition": {
+         |        "StringLike": {
+         |          "s3:prefix": [
+         |            "${path}/${uploadId}/*"
+         |          ]
+         |        }
+         |      }
+         |    },
+         |    {
+         |      "Effect": "Allow",
+         |      "Action": [
+         |        "s3:GetObject",
+         |        "s3:PutObject",
+         |        "s3:DeleteObject"
+         |      ],
+         |      "Resource": [
+         |        "arn:aws:s3:::${bucket}/${path}/${uploadId}/*"
+         |      ]
+         |    }
+         |  ]
+         |}
+       """.stripMargin.filter(_ > ' ')
+
+    val request = new GetFederationTokenRequest(userName).withPolicy(policy)
+    val response = client.getFederationToken(request)
```

Comment:
This might be better, but there are a few things I can't tell from the Auth0 documentation:
 1. How to specify a custom prefix based on the request. The example they have shows a parameterized policy based on `saml:sub`, but in our case, we want to add a prefix based on upload ID. Looks like that might be possible if we write a custom rule, but that's not particularly clear. We could forego this, but it seems like uploads could turn into a bit of a mess eventually.
 2. Do we still need the server to interact with Auth0? The example image suggests probably, and we probably still want one anyways so that our API clients do not have to interact with Auth0 and gives us an abstraction to move away if we need to.
![ex](https://cloud.githubusercontent.com/assets/898060/24295279/5f14b338-1070-11e7-80da-11574aa8a449.png)